storage/src/main/java/io/delta/storage/HadoopFileSystemLogStore.java

```diff
@@ -68,6 +70,72 @@ public Path resolvePathOnPhysicalStorage(
         return path.getFileSystem(hadoopConf).makeQualified(path);
     }
 
+    /**
+     * An internal write implementation that uses FileSystem.rename().
+     * <p>
+     * This implementation should only be used for the underlying file systems that support atomic
+     * renames, e.g., Azure is OK but HDFS is not.
+     */
+    protected void writeWithRename(
+            Path path,
+            Iterator<String> actions,
+            boolean overwrite,
+            Configuration hadoopConf) throws IOException {
+        FileSystem fs = path.getFileSystem(hadoopConf);
+
+        if (!fs.exists(path.getParent())) {
+            throw new FileNotFoundException(
+                    String.format("No such file or directory: %s", path.getParent())
+            );
+        }
+        if (overwrite) {
+            final FSDataOutputStream stream = fs.create(path, true);
+            try {
+                while (actions.hasNext()) {
+                    stream.write((actions.next() + "\n").getBytes(StandardCharsets.UTF_8));
+                }
+            } finally {
+                stream.close();
+            }
+        } else {
+            if (fs.exists(path)) {
+                throw new FileAlreadyExistsException(path.toString());
+            }
+            Path tempPath = createTempPath(path);
+            boolean streamClosed = false; // This flag is to avoid double close
+            boolean renameDone = false;   // This flag is to save the delete operation in most of cases.
+            final FSDataOutputStream stream = fs.create(tempPath);
+            try {
+                while (actions.hasNext()) {
+                    stream.write((actions.next() + "\n").getBytes(StandardCharsets.UTF_8));
+                }
+                stream.close();
+                streamClosed = true;
+                try {
+                    if (fs.rename(tempPath, path)) {
+                        renameDone = true;
+                    } else {
+                        if (fs.exists(path)) {
+                            throw new FileAlreadyExistsException(path.toString());
+                        } else {
+                            throw new IllegalStateException(
+                                    String.format("Cannot rename %s to %s", tempPath, path)
+                            );
+                        }
+                    }
+                } catch (FileAlreadyExistsException e) {
```

Comment:
catch `org.apache.hadoop.fs.FileAlreadyExistsException`.

then we will rethrow `FileAlreadyExistsException` which is actually a `java.nio.file.FileAlreadyExistsException`

Reply:
Thanks, good catch.